decode_test.go

```diff
@@ -29,601 +29,607 @@ func TestDecoder(t *testing.T) {
-			`"1": "2"`,
-			map[interface{}]interface{}{"1": `2`},
+			source: `"1": "2"`,
+			value:  map[interface{}]interface{}{"1": `2`},
 		},
 		{
-			`"1": "\"2\""`,
-			map[interface{}]interface{}{"1": `"2"`},
+			source: `"1": "\"2\""`,
+			value:  map[interface{}]interface{}{"1": `"2"`},
 		},
 		{
-			`"1": "\""`,
-			map[interface{}]interface{}{"1": `"`},
+			source: `"1": "\""`,
+			value:  map[interface{}]interface{}{"1": `"`},
 		},
 		{
-			`"1": "X\"z"`,
-			map[interface{}]interface{}{"1": `X"z`},
+			source: `"1": "X\"z"`,
+			value:  map[interface{}]interface{}{"1": `X"z`},
 		},
 		{
-			`"1": "\\"`,
-			map[interface{}]interface{}{"1": `\`},
+			source: `"1": "\\"`,
+			value:  map[interface{}]interface{}{"1": `\`},
 		},
 		{
-			`"1": "\\\\"`,
-			map[interface{}]interface{}{"1": `\\`},
+			source: `"1": "\\\\"`,
+			value:  map[interface{}]interface{}{"1": `\\`},
 		},
 		{
-			`"1": "\\\"2\\\""`,
-			map[interface{}]interface{}{"1": `\"2\"`},
+			source: `"1": "\\\"2\\\""`,
+			value:  map[interface{}]interface{}{"1": `\"2\"`},
 		},
 		{
-			"'1': \"   1\n    2\n    3\"",
-			map[interface{}]interface{}{"1": "   1 2 3"},
+			source: "'1': \"   1\n    2\n    3\"",
+			value:  map[interface{}]interface{}{"1": "   1 2 3"},
 		},
 		{
-			"'1': \"\n    2\n    3\"",
-			map[interface{}]interface{}{"1": " 2 3"},
+			source: "'1': \"\n    2\n    3\"",
+			value:  map[interface{}]interface{}{"1": " 2 3"},
 		},
 		{
-			`"1": "a\x2Fb"`,
-			map[interface{}]interface{}{"1": `a/b`},
+			source: `"1": "a\x2Fb"`,
+			value:  map[interface{}]interface{}{"1": `a/b`},
 		},
 		{
-			`"1": "a\u002Fb"`,
-			map[interface{}]interface{}{"1": `a/b`},
+			source: `"1": "a\u002Fb"`,
+			value:  map[interface{}]interface{}{"1": `a/b`},
 		},
 		{
-			`"1": "a\x2Fb\u002Fc\U0000002Fd"`,
-			map[interface{}]interface{}{"1": `a/b/c/d`},
+			source: `"1": "a\x2Fb\u002Fc\U0000002Fd"`,
+			value:  map[interface{}]interface{}{"1": `a/b/c/d`},
 		},
 		{
-			"'1': \"2\\n3\"",
-			map[interface{}]interface{}{"1": "2\n3"},
+			source: "'1': \"2\\n3\"",
+			value:  map[interface{}]interface{}{"1": "2\n3"},
 		},
 		{
-			"'1': \"2\\r\\n3\"",
-			map[interface{}]interface{}{"1": "2\r\n3"},
+			source: "'1': \"2\\r\\n3\"",
+			value:  map[interface{}]interface{}{"1": "2\r\n3"},
 		},
 
 		{
-			"a: -b_c",
-			map[string]interface{}{"a": "-b_c"},
+			source: "a: -b_c",
+			value:  map[string]interface{}{"a": "-b_c"},
 		},
 		{
-			"a: +b_c",
-			map[string]interface{}{"a": "+b_c"},
+			source: "a: +b_c",
+			value:  map[string]interface{}{"a": "+b_c"},
 		},
 		{
-			"a: 50cent_of_dollar",
-			map[string]interface{}{"a": "50cent_of_dollar"},
+			source: "a: 50cent_of_dollar",
+			value:  map[string]interface{}{"a": "50cent_of_dollar"},
 		},
 
 		// Nulls
 		{
-			"v:",
-			map[string]interface{}{"v": nil},
+			source: "null",
+			value:  (*struct{})(nil),
 		},
 		{
-			"v: ~",
-			map[string]interface{}{"v": nil},
+			source: "~",
+			value:  (*struct{})(nil),
```

Comment:
Sorry for the large diff 🙇 This is the only new test case I added. I wanted to distinguish the io.EOF error from others, so I added an eof field to the test struct.